Authentication Updates pt. 2

salvis2 opened the pull request:
Opening a new PR for authenticator updates. Still uses context managers. I found some bugs in the old version, which I have since fixed. These updates allowed me to deploy a hub locally and through GitHub Actions, see [here](https://github.com/salvis2/hubploy/actions/runs/50408017).

The commit history is different than the other PR, since I am in the middle of fixing my git workflow and don't want a messy merge in this PR's history.

yuvipanda commented on hubploy/auth.py:
```diff
@@ -182,24 +199,33 @@ def cluster_auth_gcloud(deployment, project, cluster, zone, service_key):
+    os.environ["AWS_SHARED_CREDENTIALS_FILE"] = service_key_path
+
+    try:
+        subprocess.check_call(['aws', 'eks', 'update-kubeconfig',
```
Actually, what we care about here is the correct path to the kubernetes config, without modifying the user's current kubernetes context. To that end, I propose the following:

1. Get a secure temporary file name from tempfile module
2. Set the `KUBECONFIG` environment variable to the temporary file name. 
3. Run the `update-kubeconfig` command. This will actually set up the file in the path `KUBECONFIG` points to.
4. All other commands (helm, etc) will use the path specified in `KUBECONFIG`
5. In our finally, we can unset `KUBECONFIG`, and delete the file.

This technique will actually work for google cloud (tested), and probably Azure too. 

What do you think of this?

yuvipanda replied:
We should also save the current value of `KUBECONFIG` and restore it when done.

salvis2 replied:
Setting and unsetting the `KUBECONFIG` seems to be the same pattern we are using for the AWS credentials (which also avoids changing the user's settings, as you said), so I think that's a good idea. Keep everything working the same way for clarity (and stop messing with user files).

Seems like a straightforward set of changes, I'll work on it.

salvis2 replied:
Ah. The main problem I'm anticipating is that if I wanted to test and manage hubploy and the cluster locally, I'd need a manual `aws eks update-kubeconfig` call to do that, since the config location is reverted once hubploy is done working. That isn't a terribly difficult command to expect someone to run, though.

yuvipanda replied:
@salvis2 yeah, you'd need to explicitly do that - this will leave your context unchanged. I think that's the 'right' behavior though.

yuvipanda replied:
Yeah, I think that is the right thing to do - people will need to explicitly call that to change their context.

salvis2 commented on hubploy/auth.py:
```diff
@@ -72,28 +81,33 @@ def registry_auth_aws(deployment, project, zone, service_key):
-        json.dump(config, f)
 
+    # Set env variable for credential file location
+    os.environ["AWS_SHARED_CREDENTIALS_FILE"] = service_key_path
```
Okay, that should be a pretty easy change. 

I'm also fuzzy on the `try` block, but it's probably cleaner to have all the code inside `try` or `finally`.

salvis2 commented:
Running hubploy commands worked locally for build and deploy both, but making the hubploy image and using it to deploy a JupyterHub is having a couple issues. Working on that more tomorrow.

yuvipanda commented on hubploy/auth.py:
```diff
@@ -182,24 +201,47 @@ def cluster_auth_gcloud(deployment, project, cluster, zone, service_key):
+    finally:
+        # Unset env variable for credential file location
+        del os.environ["AWS_SHARED_CREDENTIALS_FILE"]
+        os.environ["AWS_SHARED_CREDENTIALS_FILE"] = original_credential_file_loc
```
If the env variable was unset originally, we should leave it unset - rather than setting it to "". Apps often change behavior on wether the variable is set or not, regardless of the value.

salvis2 commented:
Tried to clean up the last couple things you mentioned. Thank you!

Main problem I will detail now (since I am stumped): running `hubploy build ...` and `hubploy deploy ...` on my own machine with these commits works great. I can build the hubploy image and push it to dockerhub under `sebialvis/hubploy`, I update that name in my `hubploy-template` [fork](https://github.com/salvis2/hackweek-template/blob/staging/.github/workflows/deploy.yaml) and push to staging.

The GitHub Action [fails](https://github.com/salvis2/hackweek-template/runs/565294537?check_suite_focus=true) with the following output.

```
Added new context arn:aws:eks:us-west-2:783380859522:cluster/hackweek-cluster to /tmp/tmp77styap0
Saving 1 charts
Downloading jupyterhub from repo https://jupyterhub.github.io/helm-chart
Deleting outdated charts
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/hubploy/helm.py", line 56, in helm_upgrade
    kubernetes.config.load_kube_config()
  File "/usr/local/lib/python3.7/site-packages/kubernetes/config/kube_config.py", line 739, in load_kube_config
    persist_config=persist_config)
  File "/usr/local/lib/python3.7/site-packages/kubernetes/config/kube_config.py", line 701, in _get_kube_config_loader_for_yaml_file
    'Invalid kube-config file. '
kubernetes.config.config_exception.ConfigException: Invalid kube-config file. No configuration found.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/hubploy", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.7/site-packages/hubploy/__main__.py", line 87, in main
    helm.deploy(args.deployment, args.chart, args.environment, args.namespace, args.set, args.version, args.timeout, args.force)
  File "/usr/local/lib/python3.7/site-packages/hubploy/helm.py", line 151, in deploy
    force
  File "/usr/local/lib/python3.7/site-packages/hubploy/helm.py", line 58, in helm_upgrade
    kubernetes.config.load_incluster_config()
  File "/usr/local/lib/python3.7/site-packages/kubernetes/config/incluster_config.py", line 94, in load_incluster_config
    cert_filename=SERVICE_CERT_FILENAME).load_and_set()
  File "/usr/local/lib/python3.7/site-packages/kubernetes/config/incluster_config.py", line 45, in load_and_set
    self._load_config()
  File "/usr/local/lib/python3.7/site-packages/kubernetes/config/incluster_config.py", line 51, in _load_config
    raise ConfigException("Service host/port is not set.")
kubernetes.config.config_exception.ConfigException: Service host/port is not set.
```

I think having two errors is just the result of [this](https://github.com/yuvipanda/hubploy/blob/master/hubploy/helm.py#L57) block, and only the first one indicates anything that might be useful.

Unsure why the kubeconfig file isn't read on the CI machine, since it works locally. Also strange because the `aws eks update-kubeconfig ...` command seems to work.

Any thoughts?

yuvipanda commented:
I haven't looked at the code yet, but can you check if your local ~/.kube/config was empty / deleted before running this? Otherwise it might just be picking it up from there. It could also be because KUBECONFIG is set  to "" at some point - it should be unset, not set to "" after it's done. In fact, we should never set any env vars to "" - we should probably use `None` as a default.

I'll try look at the code later today though!

salvis2 commented:
To your second comment, I've cleaned up how I'm unsetting / resetting the environment variables. It's in the last commit.

To the first comment, I emptied my local `~/.kube/config` file and tried deploying, and the error behavior was almost exactly as above. The first error message was different, but the second was the same.

```
Added new context arn:aws:eks:us-west-2:783380859522:cluster/hackweek-cluster to /tmp/tmp8fdmbsdj
Hang tight while we grab the latest from your chart repositories...
...Successfully got an update from the "dask" chart repository
...Successfully got an update from the "stable" chart repository
Update Complete. ⎈Happy Helming!⎈
Saving 1 charts
Downloading jupyterhub from repo https://jupyterhub.github.io/helm-chart
Deleting outdated charts
Traceback (most recent call last):
  File "/home/salvis/Documents/GitHub/hubploy/hubploy/helm.py", line 58, in helm_upgrade
    kubernetes.config.load_kube_config()
  File "/home/salvis/Documents/GitHub/hackweek-template/lib/python3.6/site-packages/kubernetes/config/kube_config.py", line 739, in load_kube_config
    persist_config=persist_config)
  File "/home/salvis/Documents/GitHub/hackweek-template/lib/python3.6/site-packages/kubernetes/config/kube_config.py", line 695, in _get_kube_config_loader_for_yaml_file
    kcfg = KubeConfigMerger(filename)
  File "/home/salvis/Documents/GitHub/hackweek-template/lib/python3.6/site-packages/kubernetes/config/kube_config.py", line 650, in __init__
    self.load_config(path)
  File "/home/salvis/Documents/GitHub/hackweek-template/lib/python3.6/site-packages/kubernetes/config/kube_config.py", line 664, in load_config
    config_merged[item] = []
TypeError: 'NoneType' object does not support item assignment

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/salvis/Documents/GitHub/hackweek-template/bin/hubploy", line 11, in <module>
    load_entry_point('hubploy', 'console_scripts', 'hubploy')()
  File "/home/salvis/Documents/GitHub/hubploy/hubploy/__main__.py", line 105, in main
    helm.deploy(args.deployment, args.chart, args.environment, args.namespace, args.set, args.version, args.timeout, args.force, args.atomic, args.cleanup_on_fail)
  File "/home/salvis/Documents/GitHub/hubploy/hubploy/helm.py", line 161, in deploy
    cleanup_on_fail
  File "/home/salvis/Documents/GitHub/hubploy/hubploy/helm.py", line 60, in helm_upgrade
    kubernetes.config.load_incluster_config()
  File "/home/salvis/Documents/GitHub/hackweek-template/lib/python3.6/site-packages/kubernetes/config/incluster_config.py", line 94, in load_incluster_config
    cert_filename=SERVICE_CERT_FILENAME).load_and_set()
  File "/home/salvis/Documents/GitHub/hackweek-template/lib/python3.6/site-packages/kubernetes/config/incluster_config.py", line 45, in load_and_set
    self._load_config()
  File "/home/salvis/Documents/GitHub/hackweek-template/lib/python3.6/site-packages/kubernetes/config/incluster_config.py", line 51, in _load_config
    raise ConfigException("Service host/port is not set.")
kubernetes.config.config_exception.ConfigException: Service host/port is not set.
```

A great suggestion! I will poke more into how to fix this locally, and then I can test remotely and push some more commits.

yuvipanda commented on hubploy/auth.py:
```diff
@@ -72,28 +82,33 @@ def registry_auth_aws(deployment, project, zone, service_key):
-            config = json.load(f)
-    else:
-        config = {}
+    original_credential_file_loc = os.environ.get("AWS_SHARED_CREDENTIALS_FILE", "")
```
Let's use `None` as the default, not ""? IMO that's good practice for fiddling with env vars temporarily - we *do* want to differentiate between empty value "" and a marker that it did not exist

salvis2 replied:
Okay. I can change that and the unsetting logic so that it looks for `None`.

yuvipanda commented on hubploy/auth.py:
```diff
@@ -72,28 +82,33 @@ def registry_auth_aws(deployment, project, zone, service_key):
 
+    finally:
+        # Unset env variable for credential file location
+        unset_env_var("AWS_SHARED_CREDENTIALS_FILE", original_credential_file_loc)
```
IMO this function is small enough that we can just write it out here:

```python
del os.environ["AWS_SHARED_CREDENTIALS_FILE"]
if original_credential_file_loc is not None:
	os.environ["AWS_SHARED_CREDENTIALS_FILE"] = original_credential_file_loc
```

salvis2 replied:
Well, I have to use that function two more times in `cluster_auth_aws()`, so I feel like it's easier to leave it as a function. I can change the logic to match this one, since deleting the value hubploy used seems like a good thing to do always.

yuvipanda replied:
Fair!

salvis2 commented:
Testing the Github Actions functionality...

salvis2 commented:
Worked locally.

Successfully built hubploy [here](https://github.com/salvis2/hubploy/actions/runs/73727555).

Successfully deployed hub through GitHub [Action](https://github.com/salvis2/hackweek-template/actions/runs/73732894).

yuvipanda commented:
w00p w00p, thanks @salvis2! I just tested this locally on google cloud, and it worked fine. \o/ Thank you so much!
